atomix/cluster/src/test/java/io/atomix/raft/storage/log/RaftLogTest.java

```diff
@@ -0,0 +1,166 @@
+import io.atomix.raft.cluster.impl.DefaultRaftMember;
+import io.atomix.raft.storage.log.entry.ApplicationEntry;
+import io.atomix.raft.storage.log.entry.ConfigurationEntry;
+import io.atomix.raft.storage.log.entry.InitialEntry;
+import io.atomix.raft.storage.log.entry.RaftLogEntry;
+import java.io.File;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.time.Instant;
+import java.util.Set;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.io.TempDir;
+
+class RaftLogTest {
+  @TempDir File directory;
+
+  private final InitialEntry initialEntry = new InitialEntry();
+  private final ConfigurationEntry configurationEntry =
+      new ConfigurationEntry(
+          1234L,
+          Set.of(
+              new DefaultRaftMember(MemberId.from("0"), Type.ACTIVE, Instant.ofEpochSecond(1234))));
+  private final ByteBuffer data = ByteBuffer.allocate(8);
+  private final ApplicationEntry applicationEntry = new ApplicationEntry(1, 2, data);
+  private RaftLog raftlog;
+  private RaftLogReader reader;
+
+  @BeforeEach
+  void setup() {
+    raftlog = RaftLog.builder().withDirectory(directory).withName("test").build();
+    reader = raftlog.openReader();
+    data.order(ByteOrder.LITTLE_ENDIAN).putInt(123456);
+  }
+
+  @AfterEach
+  void tearDown() {
+    reader.close();
+    raftlog.close();
+  }
+
+  @Test
+  void shouldAppendInitialEntry() {
+    // given
+    final RaftLogEntry entry = new RaftLogEntry(1, initialEntry);
+    // when
+    final var appended = raftlog.append(entry);
+
+    // then
+    assertThat(reader.hasNext()).isTrue();
+    final var entryRead = reader.next();
+    assertThat(entryRead.term()).isEqualTo(1);
+    assertThat(entryRead.entry()).isInstanceOf(InitialEntry.class);
+    assertThat(appended).isEqualTo(entryRead);
+  }
+
+  @Test
+  void shouldAppendConfigurationEntry() {
+    // given
+    final RaftLogEntry entry = new RaftLogEntry(1, configurationEntry);
+    // when
+    final var appended = raftlog.append(entry);
+
+    // then
+    assertThat(reader.hasNext()).isTrue();
+    final var entryRead = reader.next();
+    assertThat(entryRead.term()).isEqualTo(1);
+    assertThat(entryRead.entry()).isInstanceOf(ConfigurationEntry.class);
+  }
+
+  @Test
+  void shouldAppendApplicationEntry() {
+    // given
+    final RaftLogEntry entry = new RaftLogEntry(1, applicationEntry);
+    // when
+    final var appended = raftlog.append(entry);
+
+    // then
+    assertThat(reader.hasNext()).isTrue();
+    final var entryRead = reader.next();
+    assertThat(entryRead.term()).isEqualTo(1);
+    assertThat(entryRead.isApplicationEntry()).isTrue();
+    assertThat(entryRead.getApplicationEntry().lowestPosition()).isEqualTo(1);
+    assertThat(entryRead.getApplicationEntry().highestPosition()).isEqualTo(2);
+    assertThat(entryRead.getApplicationEntry().data().getInt(0)).isEqualTo(data.getInt(0));
```

Comment:
Shouldn't we expect the buffers to be the same, regardless of contents? So, something like:

```suggestion
    assertThat(entryRead.getApplicationEntry().data()).isEqualTo(new UnsafeBuffer(data));
```